controllers/twophase/types.go

```diff
@@ -61,206 +61,73 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 	chaos := _chaos.(v1alpha1.InnerSchedulerObject)
 
-	duration, err := chaos.GetDuration()
-	if err != nil {
-		r.Log.Error(err, "failed to get chaos duration")
-		return ctrl.Result{}, err
-	}
+	status := chaos.GetStatus()
+
+	targetPhase := status.Experiment.Phase
 
-	scheduler := chaos.GetScheduler()
-	if scheduler == nil {
-		r.Log.Info("Scheduler should be defined currently")
-		return ctrl.Result{}, fmt.Errorf("misdefined scheduler")
+	if !chaos.GetNextRecover().IsZero() && chaos.GetNextRecover().Before(now) {
+		targetPhase = v1alpha1.ExperimentPhaseWaiting
 	}
 
-	if duration == nil {
-		zero := 0 * time.Second
-		duration = &zero
+	if chaos.GetNextStart().Before(now) {
+		targetPhase = v1alpha1.ExperimentPhaseRunning
 	}
 
-	status := chaos.GetStatus()
+	if chaos.IsPaused() {
+		targetPhase = v1alpha1.ExperimentPhasePaused
+	}
 
 	if chaos.IsDeleted() {
-		// This chaos was deleted
-		r.Log.Info("Removing self")
-		err = r.Recover(ctx, req, chaos)
-		if err != nil {
-			r.Log.Error(err, "failed to recover chaos")
-			updateFailedMessage(ctx, r, chaos, err.Error())
-			return ctrl.Result{Requeue: true}, err
-		}
-
-		status.Experiment.Phase = v1alpha1.ExperimentPhaseFinished
-		status.FailedMessage = emptyString
-	} else if chaos.IsPaused() {
-		if status.Experiment.Phase == v1alpha1.ExperimentPhaseRunning {
-			r.Log.Info("Pausing")
-
-			err = r.Recover(ctx, req, chaos)
-			if err != nil {
-				r.Log.Error(err, "failed to pause chaos")
-				updateFailedMessage(ctx, r, chaos, err.Error())
-				return ctrl.Result{Requeue: true}, err
-			}
-
-			now := time.Now()
-			status.Experiment.EndTime = &metav1.Time{
-				Time: now,
-			}
-			if status.Experiment.StartTime != nil {
-				status.Experiment.Duration = now.Sub(status.Experiment.StartTime.Time).String()
-			}
-		}
-		status.Experiment.Phase = v1alpha1.ExperimentPhasePaused
-		status.FailedMessage = emptyString
-	} else if !chaos.GetNextRecover().IsZero() && chaos.GetNextRecover().Before(now) {
-		// Start recover
-		r.Log.Info("Recovering")
-
-		// Don't need to recover again if chaos was paused before
-		if status.Experiment.Phase != v1alpha1.ExperimentPhasePaused {
-			if err = r.Recover(ctx, req, chaos); err != nil {
-				r.Log.Error(err, "failed to recover chaos")
-				updateFailedMessage(ctx, r, chaos, err.Error())
-				return ctrl.Result{Requeue: true}, err
-			}
-		}
-
-		chaos.SetNextRecover(time.Time{})
-
-		status.Experiment.EndTime = &metav1.Time{
-			Time: time.Now(),
-		}
-		status.Experiment.Phase = v1alpha1.ExperimentPhaseWaiting
-		status.FailedMessage = emptyString
-	} else if (status.Experiment.Phase == v1alpha1.ExperimentPhaseFailed ||
-		status.Experiment.Phase == v1alpha1.ExperimentPhasePaused) &&
-		!chaos.GetNextRecover().IsZero() && chaos.GetNextRecover().After(now) {
```

Comment:
@YangKeao We've been missing this part check, we're still in this current round!

Reply:
Yes. It failed to move forward from a `PhasePause`... `targetPhase` was initialized as `PhasePause`, and it doesn't have enough evidence to move forward to another phase.

I have added an ugly check to decide whether it's resuming from pause.

```
if !chaos.IsPaused() && status.Experiment.Phase == v1alpha1.ExperimentPhasePaused {
	// Running and Waiting has the same logic for resuming
	targetPhase = v1alpha1.ExperimentPhaseRunning
}
```

Actually, the state machine has a lot more input than `Phase`. The `nextStart`, `nextRecover` and `lastStart` are all the inputs to resolve the pause problem... I failed to design a clean mechanism 😿 .

Reply:
The ideal state-machine implementation should be an event-driven model, rather than deciding the target phase according these `nextStart`, `nextRecover`, `lastStart `...

The  root cause why we need these if-else check, I think it is that we've been **mixed** schedule and state machine input event into the same one controller, so we cared about these chaos current state, and according these state to decide the target state.